File: 344. Reverse String/reverse_string.go
// Iterative solution
func reverseString(s string) string {
    reversed := ""

    for i, _ := range s {
        reversed += string(s[(len(s) -1) - i])
    }
    
    return reversed
}


// Recursive solution
func reverseString(s string) string {

	if s == "" {
		return s
	}

	return string(s[len(s) - 1]) + reverseString(s[:len(s) - 1])
}

File: reverse_string/reverse_string.py
# Solution for cheaters
class Solution(object):
    def reverseString(self, s):
        """
        :type s: str
        :rtype: str
        """
        return s[::-1]


# Solutions for regular fellas -- with comprehension...
class Solution(object):
    def reverseString(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = ''.join([s[-i] for i in range(1, len(s) + 1)])
        return res



# ... iterative
class Solution(object):
    def reverseString(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = ""
        
        for i in range(1, len(s) + 1):
        	res += s[-i]
        
        return res 


# ... and with tail recursion
class Solution(object):
    def reverseString(self, s):
        """
        :type s: str
        :rtype: str
        """
        def accumulate(acc, src):
            if len(src) == 0:
                return acc
            return accumulate(acc + src[len(src) - 1], src[:len(src) - 1])         


        return accumulate("", s)

